### PR TITLE
feat: upgrade langchain version

### PR DESCRIPTION
**Pull request by jczhong84**
- Upgrade langchain to 0.1.6
- Adopt the new streaming approach which support partial json streaming and remove the old streaming code
- Also add non streaming support
- Split sql edit from the text-to-sql prompt

### CODE REVIEW

File: querybook/webapp/components/AIAssistant/QueryGenerationModal.tsx [hunk 1]
```diff
@@ -103,6 +102,10 @@ export const QueryGenerationModal = ({
     const [streamData, setStreamData] = useState<{ [key: string]: string }>({});
     const [foundTables, setFoundTables] = useState<string[]>([]);
 
+    const debounceSetStreamData = debounce((data) => {
```
[the remark this answers is not in the document]
Comment (jczhong84) [reply querybook/webapp/components/AIAssistant/QueryGenerationModal.tsx:105]: good catch!

File: querybook/server/lib/ai_assistant/prompts/sql_title_prompt.py [hunk 2]
```diff
@@ -1,17 +1,19 @@
...
-    "{query}\n\n"
-    "===Response Guidelines\n"
-    "1. Only respond with the title without any explanation\n"
-    "2. Dont use double quotes to enclose the title\n"
-    "3. Dont add a final period to the title\n\n"
-    "===Example response\n"
-    "This is a title\n"
-)
+prompt_template = """
+You are a helpful data scientist that can summerize SQL queries.
```
Comment (kgopal492) [inline querybook/server/lib/ai_assistant/prompts/sql_title_prompt.py:5]: nit: summarize

**Review by kgopal492 — APPROVED** [review]

File: querybook/server/lib/ai_assistant/base_ai_assistant.py [hunk 3]
```diff
@@ -184,6 +181,26 @@ def _get_query_execution_error(self, query_execution: QueryExecution) -> str:
...
+        socket: AIWebSocket,
+        command: AICommandType,
+        llm: BaseLanguageModel,
+        prompt_text: str,
+    ):
+        """Run the prompt and send the response to the websocket. If the command is streaming, send the response in streaming mode."""
+
+        chain = llm | JsonOutputParser()
+
+        if self._get_llm_config(command.value).get("streaming", True):
```
Comment (czgu) [inline querybook/server/lib/ai_assistant/base_ai_assistant.py:195]: default streaming to False?

File: querybook/webapp/components/AIAssistant/AutoFixButton.tsx [hunk 4]
```diff
@@ -21,26 +20,31 @@ interface IProps {
...
     const {
         data: unformattedData,
         explanation,
         fix_suggestion: suggestion,
-        fixed_query: rawFixedQuery,
+        fixed_query: fixedQuery,
     } = data;
 
-    const fixedQuery = trimSQLQuery(rawFixedQuery);
+    useEffect(() => {
```
Comment (czgu) [inline querybook/webapp/components/AIAssistant/AutoFixButton.tsx:36]: could we generalize this a bit more to handle the json streaming unstability, like
```ts
function useNonEmptyState<T>(initValue: T | () => T) => {
    const [state, _setState] = useState(initValue);
    const setState = useCallback((newValOrFunc: T | (oldVal: T) => T) => {
         _setState(oldVal => {
             let newVal: T = newValOrFunc;
             if (typeof newValOrFunc === 'function')
                 newVal = newValOrFunc(oldVal);
             return newVal || oldVal;
         });     
    }, []);
    return [state, setState];
}
```

File: querybook/server/lib/ai_assistant/base_ai_assistant.py [hunk 5]
```diff
@@ -184,6 +181,26 @@ def _get_query_execution_error(self, query_execution: QueryExecution) -> str:
...
+            for s in chain.stream(prompt_text):
+                socket.send_data(s)
+            socket.close()
+        else:
+            response = llm.invoke(prompt_text)
```
[the remark this answers is not in the document]
Comment (jczhong84) [reply querybook/server/lib/ai_assistant/base_ai_assistant.py:200]: good catch, thanks!

**Review by czgu — APPROVED** [review]

**Pull request merged**